dipy/reconst/tests/test_mapmri.py

```diff
@@ -1,50 +1,161 @@
 import numpy as np
 from dipy.data import get_gtab_taiwan_dsi
 from numpy.testing import (assert_almost_equal,
+                           assert_array_almost_equal,
                            assert_equal,
                            run_module_suite)
-from dipy.reconst.mapmri import MapmriModel, mapmri_index_matrix, mapmri_EAP
+from dipy.reconst.mapmri import MapmriModel, mapmri_index_matrix
+from dipy.reconst import dti, mapmri
 from dipy.sims.voxel import (MultiTensor, all_tensor_evecs,  multi_tensor_pdf)
 from scipy.special import gamma
 from scipy.misc import factorial
 from dipy.data import get_sphere
+from dipy.sims.voxel import add_noise
+import scipy.integrate as integrate
+import scipy.special as special
+from scipy.special import jn
+from dipy.core.sphere_stats import angular_similarity
+from dipy.direction.peaks import gfa, peak_directions
+from dipy.reconst.tests.test_dsi import sticks_and_ball_dummies
+from dipy.core.subdivide_octahedron import create_unit_sphere
+from dipy.reconst.shm import sh_to_sf
 
 
 def int_func(n):
-    f = np.sqrt(2) * factorial(n) / float(((gamma(1 + n / 2.0))
-                                           * np.sqrt(2**(n + 1) * factorial(n))))
+    f = (
+        np.sqrt(2) * factorial(n) / float(((gamma(1 + n / 2.0)) *
+                                          np.sqrt(2 ** (n + 1) *
+                                                  factorial(n))))
+    )
     return f
 
 
-def test_mapmri_metrics():
-    gtab = get_gtab_taiwan_dsi()
-    mevals = np.array(([0.0015, 0.0003, 0.0003],
-                       [0.0015, 0.0003, 0.0003]))
-    angl = [(0, 0), (60, 0)]
+def generate_signal_crossing(gtab, lambda1, lambda2, lambda3, angle2=60):
+    mevals = np.array(([lambda1, lambda2, lambda3],
+                       [lambda1, lambda2, lambda3]))
+    angl = [(0, 0), (angle2, 0)]
     S, sticks = MultiTensor(gtab, mevals, S0=100.0, angles=angl,
                             fractions=[50, 50], snr=None)
+    return S, sticks
+
+
+def callaghan_perpendicular(q, R):
+    return (2 * jn(1, 2 * np.pi * q * R)) ** 2 / (2 * np.pi * q * R) ** 2
+
+
+def gaussian_parallel(q, tau, D=0.7e-3):
+    return np.exp(-(2 * np.pi * q) ** 2 * tau * D)
+
+
+def signal_cylinder_callaghan(qvec, tau, R):
+    # assumes cylinder axis is aligned with qx
+    q_norm = np.sqrt(np.einsum('ij,ij->i', qvec, qvec))
+    q_perp = np.sqrt(np.einsum('ij,ij->i', qvec[:, 1:], qvec[:, 1:]))
+    q_par = qvec[:, 0]
+    E = callaghan_perpendicular(q_perp, R) * gaussian_parallel(q_par, tau)
+    E[q_norm == 0.] = 1.
+    return E
```

Comment:
These functions could be really useful elsewhere. Maybe put them in the `sims` module, instead of being "buried" in here?